### DISCUSSION TITLE
Can I publish a npm package using raw curl or other raw request?

### DISCUSSION THREAD

**Discussion by martibutler** [community/community#145855] [npm] [answered]
Hello

I am setting up a remote npm repository using Jfrog Artifactory.  My intention is to then forward on and publish the package from Artifactory to the public NPM registry using curl or python request library with an auth token.  I am attempting to POST my zipped up package onto https://registry.npmjs.org - is this even possible please?  Or do you always have to use the NPM cli?  Is there any documentation covering this please?

Many thanks

**Comment by monishcm** [comment]
CLI is the supported approach. Why do you want to publish via curl? 

I think, it is easier to invoke npm cli from python and publish it, as the cli does lot of pre-processing before uploading the package. Or may be depend on CI/CD pipelines to automate it, where you can invoke the cli in the automation flow.

**Comment by kriti-raj** [comment]
While the npm CLI is the easiest way to publish packages, you can indeed use curl or Python’s requests to interact directly with the npm registry. You need to ensure that the request is correctly formatted, with appropriate headers and package content. If you're planning on automating this, it’s essential to handle error responses and ensure your authentication is secure.

**Reply by martibutler** [reply]
HI @kriti-raj ! Thanks for your response.  Are there any docs about the correct formation of the request please?  I have been attempting the request but currently getting 404 "Failed to save packument".

**Reply by martibutler** [reply]
Getting 503 Service Unavailbale when sending a PUT request to https://registry.npmjs.org (also tried https://registry.npmjs.org/PACKAGE_NAME) with my Bearer token attached to the package I am publishing.  Request looks like this (obviously replaced the sensitive bits)...  Would really appreciate any help on this please...

```

Header: Authorization: Bearer AUTH_TOKEN
Method: PUT
Host: https://registry.npmjs.org
Content-type: application/json
Data: {\"_id\": \"PACKAGE_NAME\", \"name\": \"PACKAGE_NAME\", \"dist-tags\": {}, \"versions\": {\"1.0.1\": {\"name\": \"martifloyd\", \"description\": \"Test package for npm publish\", \"version\": \"1.0.1\", \"author\": \"AUTHOR_NAME\", \"main\": \"index.js\", \"devDependencies\": {}, \"license\": \"ISC\", \"directories\": {\"test\": \"test\"}, \"scripts\": {\"test\": \"echo 'hello'\"}, \"_id\": \"PACKAGE_NAME@1.0.1\", \"dist\": {\"tarball\": \"tarball_source\", \"shasum\": \"SHA_SUM\", \"integrity\": \"sha512-INTEGRITY\"}}}, \"access\": null, \"_attachments\": {\"PACKAGE_NAME-1.0.1.tgz\": {\"content_type\": \"application/octet-stream\", \"data\": \"BASE64_ENCODED_STR\", \"length\": LENGTH_OF_ENCODED_STR}}}
```

**Reply by martibutler** [reply, marked as answer]
It's fixed! Just had to fiddle about with the JSON object!
